### PR TITLE
Specs and tests for SLT and SGT opcodes

### PR DESCRIPTION
**Pull request by roynalnaruto**
This PR adds specs for the Signed Comparator Gadget (opcodes SLT and SGT) #12

### CODE REVIEW

File: src/zkevm_specs/evm/execution/slt_sgt.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    a = instruction.stack_pop()
+    b = instruction.stack_pop()
+    c = instruction.stack_push()
+
+    a8s = u256_to_u8s(U256(a))
+    b8s = u256_to_u8s(U256(b))
+    c8s = u256_to_u8s(U256(c))
+
+    # if both a and b are unsigned
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/slt_sgt.py:19]:
```suggestion
    # if both a and b are positive
```

File: src/zkevm_specs/evm/execution/slt_sgt.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    if a8s[0] < 128 and b8s[0] < 128:
+        instruction.constrain_equal(
+            instruction.select(is_sgt, b < a, a < b),
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/slt_sgt.py:22]: You cannot directly compare `b` and `a` as they are in the Fp field. You need to first use `rlc_to_bytes` to convert them back to u8 sequence and write constraints using u8s.

File: src/zkevm_specs/opcode/slt_sgt.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+from typing import Sequence
```
Comment (DreamWuGit) [inline src/zkevm_specs/opcode/slt_sgt.py:1]: only keep one file `src/zkevm_specs/evm/execution/slt_sgt.py` will be ok, this file can be removed

File: specs/opcode/12SLT_13SGT.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+	- `result` is the new top of the stack
+
+## Exceptions
+
+1. Stack underflow: `1023 <= stack pointer <= 1024`
+2. Out of gas: gas left < 3
+
+## Code
+
+See [`slt_sgt.py`](../../src/zkevm_specs/evm/execution/slt_sgt.py)
```
Comment (icemelon) [inline specs/opcode/12SLT_13SGT.md:57]:
```suggestion
See [`slt_sgt.py`](src/zkevm_specs/evm/execution/slt_sgt.py)
```
Probably just use the absolute path?

File: src/zkevm_specs/evm/execution/slt_sgt.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    b_lo = int.from_bytes(b8s[16:], "big")
+
+    a_lt_b = (a_hi < b_hi) or ((a_hi == b_hi) and (a_lo < b_lo))
+    b_lt_a = (b_hi < a_hi) or ((b_hi == a_hi) and (b_lo < a_lo))
+
+    # c should be boolean, i.e. binary result of (a < b)
+    instruction.constrain_bool(cc)
+
+    # a >=0 and b < 0
+    if a8s[0] < 128 and b8s[0] >= 128:
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/slt_sgt.py:33]: Should be `a8s[31]` and `b8s[31]`?

File: src/zkevm_specs/evm/execution/slt_sgt.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    b = instruction.stack_pop()
+    c = instruction.stack_push()
+
+    # decode RLC to bytes for a and b
+    a8s = instruction.rlc_to_bytes(a, 32)
+    b8s = instruction.rlc_to_bytes(b, 32)
+    c8s = instruction.rlc_to_bytes(c, 32)
+
+    cc = int.from_bytes(c8s, "big")
+    a_hi = int.from_bytes(a8s[:16], "big")
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/slt_sgt.py:21]: It's big-endian, right? `a8s[:16]` should be `a_lo`

File: src/zkevm_specs/evm/execution/slt_sgt.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    b8s.reverse()
+    c8s.reverse()
+
+    cc = int.from_bytes(c8s, "little")
+    a_hi = int.from_bytes(a8s[16:32], "little")
+    a_lo = int.from_bytes(a8s[0:16], "little")
+    b_hi = int.from_bytes(b8s[16:32], "little")
+    b_lo = int.from_bytes(b8s[0:16], "little")
+
+    a_lt_b = (a_hi < b_hi) or ((a_hi == b_hi) and (a_lo < b_lo))
```
Comment (DreamWuGit) [inline src/zkevm_specs/evm/execution/slt_sgt.py:33]: Just one confirm:  if a =0b10.......1 (-1), b = 0b10....0  (1) in little endian form,   does the `a_hi > b_hi ` by `a_hi = int.from_bytes(a8s[16:32],  b_hi = int.from_bytes(b8s[16:32], "little")` "little")?

Others good to me !

Comment (roynalnaruto) [reply src/zkevm_specs/evm/execution/slt_sgt.py:33]: `a` and `b` are both in the two's complement form. So `-1` would be: `0b1111...1111` and `1` would be `0b1000...0000` (both little endian representations).

And if `a < 0` and `b >= 0`, we return `a < b -> 1`. We compare `a_hi`, `b_hi`, `a_lo` and `b_lo` only if both `a` and `b` have the same sign, i.e. `a < 0 && b < 0` or `a >= 0 && b >= 0`

**Review by DreamWuGit — APPROVED** [review]

File: src/zkevm_specs/evm/execution/slt_sgt.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    b_lt_a = (b_hi < a_hi) or ((b_hi == a_hi) and (b_lo < a_lo))
+
+    # c should be boolean, i.e. binary result of (a < b)
+    instruction.constrain_bool(cc)
+
+    # a >=0 and b < 0
```
Comment (icemelon) [inline src/zkevm_specs/evm/execution/slt_sgt.py:39]:
```suggestion
    # a >= 0 and b < 0
```

**Comment by icemelon** [comment]
Close the PR as a new PR is made in the upstream

**Pull request closed without merging**